### ISSUE TITLE
Implement Icom network cat control

### ISSUE THREAD

**Issue by mdblack98** [Hamlib/Hamlib#716]
e.g. IC-9700 has a network port for CAT control
Protocol is not CI-V and need manual for the network protocol.

**Comment by mikaelnousiainen** [comment]
@mdblack98 It seems there are three UDP ports used by the network-capable Icom rigs: IC-7610, IC-785x, IC-9700, IC-R8600 and IC-705. First of them is some "control port" that is used (at least) for some form of authentication and enabling of remote control features. The second one is "serial port" that somehow uses the CI-V protocol frames inside UDP packets. The third one is audio port that is used to transfer (PCM?) audio.

It's weird that there is no mention of these additional protocols in the manuals, but it seems some hams have implemented the protocol already in an open source app: https://github.com/nonoo/kappanhang

Maybe there is some documentation available after all.

Now only missing suitable hardware for this :)

**Comment by mikaelnousiainen** [comment]
@mdblack98 Is this something you could ask docs for from Icom? They must have something. Not really sure what it is called: UDP network protocol / RS-BA1 server protocol?

**Comment by mdblack98** [comment]
I did ask...

 

    On Thursday, June 3, 2021, 02:37:59 PM CDT, Mikael Nousiainen ***@***.***> wrote:  
 
 


@mdblack98 Is this something you could ask docs for from Icom? They must have something. Not really sure what it is called: UDP network protocol / RS-BA1 server protocol?

**Comment by mdblack98** [comment]
Icom's inadequate response 

Other than the CI-V command manual that is available on the ICOM America Knowledge Base there is no other information available for radio control. The link for this information is:  https://www.dropbox.com/s/fe7wf3icqzoueko/IC-9700_CI-V%20Reference%20Guide.pdf?dl=0
 
Engineering has not provided any information to us on the network interface.
 
 Regards,
 
Fred Varian
IA Technical Support

**Comment by arodland** [comment]
Are you 100% sure it's not CI-V? I've used `socat` on a Linux machine to connect a TCP listen socket to a radio's CI-V port, then told RS-BA1 to connect to that port, and it worked.

Yes, there's some other stuff, including a discovery protocol over UDP, and a separate protocol for the remote audio, but I believe that the CAT aspect of it is 100% just CI-V over TCP.

**Comment by mdblack98** [comment]
As I recall we tried a TCP connection but not UDP.
Hamlib does not have a UDP interface so perhaps could add that and give it a try.

**Comment by mikaelnousiainen** [comment]
Like I've mentioned above, the Kappanhang project has a working implementation of the Icom network protocol.

The protocol seems to use separate UDP ports for different streams (control/auth, serial, audio) and the CI-V commands are wrapped inside UDP frames:

Control: https://github.com/nonoo/kappanhang/blob/main/controlstream.go#L41

Serial: https://github.com/nonoo/kappanhang/blob/main/serialstream.go#L33

In a way Kappanhang could used *with* (Linux only, of course) Hamlib as it seems to use loopback serial and audio devices to expose the radio connection.

This could be implemented in Hamlib too, but it requires quite a lot of logic -- and to be useful Hamlib should probably have a way to transfer audio to and from the device. Maybe through raw UDP audio streams?

**Comment by mdblack98** [comment]
This does require the RS_BA software to be running, right?
Or does the rig itself does support this?
Mike W9MDB

 

    On Tuesday, December 21, 2021, 12:30:43 AM CST, Mikael Nousiainen ***@***.***> wrote:  
 
 


Like I've mentioned above, the Kappanhang project has a working implementation of the Icom network protocol.

The protocol seems to use separate UDP ports for different streams (control/auth, serial, audio) and the CI-V commands are wrapped inside UDP frames:

Control: https://github.com/nonoo/kappanhang/blob/main/controlstream.go#L41

Serial: https://github.com/nonoo/kappanhang/blob/main/serialstream.go#L33

In a way Kappanhang could used with (Linux only, of course) Hamlib as it seems to use loopback serial and audio devices to expose the radio connection.

This could be implemented in Hamlib too, but it requires quite a lot of logic -- and to be useful Hamlib should probably have a way to transfer audio to and from the device. Maybe through raw UDP audio streams?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by arodland** [comment]
The rigs that have built-in Ethernet/WiFi have the equivalent of the RS-BA1
server onboard.

**Comment by mikaelnousiainen** [comment]
Yeah, indeed, the Icom rigs with an Ethernet port are running the UDP networking internally (you can configure ports and all in the menus), so no 3rd party software is necessary.

I guess the RS-BA1 server is meant for enabling remote control for rigs with USB/serial port only (like my IC-7300).

**Comment by guitarpicva** [comment]
kappenhang's github work is in golang which makes it a bit challenging for C/C++ folk to translate but you can derive the general sequence and format of the command and response frames to set up a connection to the server in the radio.

Thus far, I have gotten most of the way to creating the actual "login" to the radio.  To the point of initiating the UDP streams.  Had to set it aside for now.

**Comment by mikaelnousiainen** [comment]
Will be implemented as a part of ARDC Hamlib improvement project in #1942
